### ISSUE TITLE
Realizar identificação de parentesco

### ISSUE THREAD

**Issue by vitalwarley** [vitalwarley/research#33]

**Comment by vitalwarley** [comment]
Permiti que o modelo usasse todo o dataset a cada época e o deixei treinando por 20 épocas na RIG2. Resultados abaixo são insatisfatórios, eu penso.

![plot_v3](https://github.com/vitalwarley/research/assets/6365065/9ccb20b5-6fe9-4529-9b04-b1f2bf39f995)

**Comment by vitalwarley** [comment]
Havia inserido uma camada dentro da sequência de classificação

```python
        self.classification = nn.Sequential(
            torch.nn.BatchNorm1d(256),
            torch.nn.ReLU(),
            torch.nn.Linear(256, 128),
            torch.nn.BatchNorm1d(128),
            torch.nn.ReLU(),
            torch.nn.Linear(128, 11),  # number of kin relations
        )
```

todavia os resultados não melhoraram

![plot_v4](https://github.com/vitalwarley/research/assets/6365065/516ab06a-b8d6-4576-9670-63f3a0fa825a)

O treinamento foi na RIG2, com bs=60, epochs=20, onde cada época vê o dataset. Enquanto escrevo isso, lembro que os dados de treino não são aleatórios. Vou modificar e retreinar.

**Comment by vitalwarley** [comment]
Melhorou, mas piorou logo depois.

![plot_v5](https://github.com/vitalwarley/research/assets/6365065/94d344df-172d-481f-b938-95cc57633775)

**Comment by vitalwarley** [comment]
Após a call de sexta, foi identificado os alguns problemas

- "Esqueci" o softmax após a última camada linear,  todavia `nn.CrossEntropyLoss` espera os _logits_ mesmo -- penso que eu tinha tido essa dúvida, e acabei lendo a [doc](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) para decidir. 
- Esqueci de modificar o dataset para que os pares não-parentes formassem uma classe `non_kin` -- atualmente a classe pode ser qualquer uma para qualquer par, isso é, apenas 11 classes são possíveis.

Após resolver esse segundo item, os resultados foram "piores" -- faz sentido, dado que a tarefa ficou mais difícil. Vale ressaltar que o dataset é bem desbalanceado, onde, por exemplo, há 10k amostras para o tipo de parentesco `fs`, mas menos de 1k para `gmgs`. Considerando todos os pares e tipos de relacionamentos, temos ainda cerca de 50k `non-kin`.

![plot](https://github.com/vitalwarley/research/assets/6365065/abde3793-6c4e-46d8-80f3-7fd80327d265)

**Comment by vitalwarley** [comment]
Vou tentar novamente, mas com filtrando os quatro pares menos frequentes.

![image](https://github.com/vitalwarley/research/assets/6365065/11fcb7c0-0578-4560-82f2-44d4e3f9b5e1)

**Comment by vitalwarley** [comment]
Em 12 épocas, a acurácia estagnou em <0.447.

**Comment by vitalwarley** [comment]
Após corrigir o viés acima, isso é, ajustar a saída da rede para o número de classes corretamente, eu repliquei o treinamento de identificação das classes fs, ms, md, fd. Abaixo o resultado

![20230902_100025](https://github.com/vitalwarley/research/assets/6365065/357a63f4-7ed3-455f-b6ea-cdd0b8c14430)

**Comment by vitalwarley** [comment]
Pensei que havia pares `non-kin` no set de treino, mas não há. Apenas nos sets de validação e teste.

**Comment by vitalwarley** [comment]
> Pensei que havia pares `non-kin` no set de treino, mas não há. Apenas nos sets de validação e teste.

A razão para tal está na seção III do artigo

![image](https://github.com/vitalwarley/research/assets/6365065/d0515500-1740-42f7-a357-72ac43b37a54)

> In our supervised contrastive learning framework, negative samples are collected by combining positive samples from different families. Typically, the contrastive loss makes samples belonging to the same family close to each other in the feature space, while samples belonging to different families are separated from each other. The hyper-parameter τ is used to control the degree of punishment for hard samples, and the smaller the value of τ , the greater the degree of punishment for hard samples.

https://github.com/vitalwarley/research/blob/92f57e2aa6388fc702e44aa06622bbc450d734c1/rfiw2021/Track1/losses.py#L6-L15

Durante o treinamento, `x1` e `x2` são batches contendo as imagens do indivíduo 1 e 2, respectivamente. Os autores concatenam os batches duas vezes e formam os tensores `x1x2` e `x2x1`. Em seguida, computa-se a similaridade entre todos pares em `x1x2`, como exemplificado abaixo. A soma do exponencial da matriz de similaridades ao longo das colunas (exceto nas diagonais) forma o denominador.

![image](https://github.com/vitalwarley/research/assets/6365065/3382249c-b7a7-46aa-a268-fae89dc06b39)

Posteriormente obtém-se a similaridade entre os tensores `x1x2` e `x2x1`. Dado que o batch contém diferentes famílias, isso é, há apenas uma família por par, essa similaridade é basicamente de pares negativos. O exponencial dos elementos (similaridades) de cada matriz consiste do numerador da fórmula.

A ideia geral, ao meu ver, é que a divisão convirja para de 1 a partir da esquerda tal que o negativo do log se aproxime de 0 pela direita. As similaridades para minimizar estão no numerador, logo seu valor deve tender a zero. Similarmente, as similaridades para maximizar estão no denomidador, mas lá também há similaridades para minimizar: o exponencial da similaridade dos pares positivos (1 de n) são bem maiores que zero, enquanto que o exponencial da similaridade dos pares negativos (n - 1 elementos) tende zero. Dessa forma, a soma dos exponenciais é dominada pelos pares positivos, o que maximiza o denominador. 

![image](https://github.com/vitalwarley/research/assets/6365065/73c3afa8-44c8-4c31-8823-df661e0e6153)

Agora a questão é: como inserir informações sobre a diferença de idade (e gênero, possivelmente) nessa perda?

**Comment by vitalwarley** [comment]
Na verdade, o numerador deve também ser maximizado. `x1x2` e `x2x1` contém apenas pares positivos. Considere x1 = (a1, b1, c1), x2 = (a2, b2, c3). Temos, então

```
  x1x2 x2x1  similaridade
0   a1   a2             1
1   b1   b2             1
2   c1   c1             1
3   a2   a1             1
4   b2   b1             1
5   c1   c1             1
```

Isso reflete a equação 1 do artigo.

Considerando isso, o numerador deve crescer o máximo possível. Abaixo um gráfico exemplificando o processo

![image](https://github.com/vitalwarley/research/assets/6365065/dbb99bf5-8cb0-475c-ae2c-0c90e6d9ad9c)

Esse outro gráfico mostra o que conseguimos ao variar `beta`

![image](https://github.com/vitalwarley/research/assets/6365065/cc78965d-2fcd-42b3-a85a-777d81a8895e)

**Comment by vitalwarley** [comment]
Será realizada em futuras estratégias ainda, possivelmente.
